app/lib/shared/package_memcache.dart

```diff
@@ -33,105 +32,63 @@ abstract class UIPackageCache {
 
 /// Uses a [Memache] to set/get/invalidate metadata for packages.
 class AppEnginePackageMemcache implements PackageCache, UIPackageCache {
-  final Memcache _memcache;
-
-  AppEnginePackageMemcache(this._memcache);
+  final SimpleMemcache<String> _json;
+  final SimpleMemcache<_PVKey> _uiPage;
+  final SimpleMemcache<String> _uiIndexPage;
+
+  AppEnginePackageMemcache(Memcache memcache)
+      : _json = new SimpleMemcache(
+            _logger, memcache, packageJsonPrefix, packageJsonExpiration),
+        _uiPage = new SimpleMemcache<_PVKey>(
+            _logger, memcache, packageUiPagePrefix, packageUiPageExpiration),
+        _uiIndexPage = new SimpleMemcache<String>(
+            _logger, memcache, '', indexUiPageExpiration);
 
   @override
-  Future<List<int>> getPackageData(String package) async {
-    final result = await _ignoreErrors(
-        _memcache.get(_packageKey(package), asBinary: true));
-
-    if (result != null)
-      _logger.info('memcache["$package"] found');
-    else
-      _logger.info('memcache["$package"] not found');
-
-    return result;
-  }
+  Future<List<int>> getPackageData(String package) => _json.getBytes(package);
 
   @override
-  Future setPackageData(String package, List<int> data) {
-    _logger.info('memcache["$package"] setting to new data');
-    return _ignoreErrors(_memcache.set(_packageKey(package), data,
-        expiration: packageJsonExpiration));
-  }
+  Future setPackageData(String package, List<int> data) =>
+      _json.setBytes(package, data);
 
   @override
-  Future<String> getUIPackagePage(String package, String version) async {
-    final result = await _ignoreErrors(
-        _memcache.get(_packageUIKey(package, version), asBinary: true));
-
-    if (result != null) {
-      _logger.info('memcache["$package"] rendered UI found');
-      return UTF8.decode(result);
-    }
-
-    _logger.info('memcache["$package"] rendered UI not found');
-    return null;
-  }
+  Future<String> getUIPackagePage(String package, String version) =>
+      _uiPage.getText(new _PVKey(package, version));
 
   @override
-  Future setUIPackagePage(String package, String version, String data) async {
-    _logger.info('memcache["$package"] setting to new rendered UI data');
-    return _ignoreErrors(_memcache.set(
-        _packageUIKey(package, version), UTF8.encode(data),
-        expiration: packageUiPageExpiration));
-  }
+  Future setUIPackagePage(String package, String version, String data) =>
+      _uiPage.setText(new _PVKey(package, version), data);
```

Comment:
I would prefer if we would not introduce extra classes like `_PVKey`. They provide very little value and introduce extra overhead.

The `SimpleMemcache` doesn't need to be generic. The value it provides atm is auto-prefixing and auto-expiration.

Reply:
It also masks the way how keys are generated, and through the type (if it is not `String`) it makes sure that the key generation is used consistently between `get*`, `set*` and `invalidate`.

If you want to remove the `_PVKey` we need to put the key creation somewhere else, potentially outside of the `SimpleMemcache`, because it may be 0-3 parameters that could go into the key.

Alternatively, I could make the API like the following, allowing the caller to decide how many of these params it will use
```
Future<String> getText({String package, String version, String context});
```
(`context` being the `panaVersion` for `SearchMemcache`).

I'm fine with either approach, which would you like to have?

Reply:
Just instead of relying on `new _PVKey(a, b, c).toString()` `_pvKey(a, b, c)` is enough.

The extra "safety" by making it generic is dminishing small, because there are max 3 calls to it with the same combined key, namely get/set/invalidate.

Also others just use `String` as the generic argument. You could argue they should then suddenly use `SimpleMemcache<Uri>`. Then we might need to do URI parsing and toString just for that.

Just thinking this is overkill.

Reply:
Yeah, good point with the uri. Changing it to String...